dix/req.c

```diff
@@ -1625,13 +1625,32 @@ static void dix_rop_completed(struct m0_sm_group *grp, struct m0_sm_ast *ast)
 	struct m0_dix_rop_ctx *rop_del_phase2 = NULL;
 	bool                   del_phase2 = false;
 	struct m0_dix_cas_rop *cas_rop;
+	uint32_t	       req_cnt = 0;
+	int		       rc;
+	bool		       cas_success = false;
 
 	(void)grp;
 	if (req->dr_type == DIX_NEXT)
 		m0_dix_next_result_prepare(req);
 	else
-		m0_tl_for(cas_rop, &rop->dg_cas_reqs, cas_rop) {
-			dix_cas_rop_rc_update(cas_rop, 0);
+		/*
+		 * Check return status of N cas request, if any one CAS
+		 * request is success, we return success to client. For failed
+		 * request skipping to send rc_update() but if it is last request
+		 * and we do not found any success request reply so far then
+		 * do update rc_update() for last failed CAS request also.
+		 */
+		m0_tl_for (cas_rop, &rop->dg_cas_reqs, cas_rop) {
```

Comment:
What do you think about simplifying the logic in the following way?

```c
if (m0_tl_forall(cas_rop, cas_rop,
                      &rop->dg_cas_reqs, cas_rop->crp_creq.ccr_sm.sm_rc != 0))
    dix_cas_rop_rc_update(cas_rop_tlist_tail(&rop->dg_cas_reqs), 0); 

m0_tl_for (cas_rop, &rop->dg_cas_reqs, cas_rop) {
  if (cas_rop->crp_creq.ccr_sm.sm_rc == 0)
    dix_cas_rop_rc_update(cas_rop, 0);
  m0_cas_req_fini(&cas_rop->crp_creq);
} m0_tl_endfor;
```

Reply:
This logic also looks good to me as it is eliminating extra 3 variable I used (rc, req_cnt, cas_success)  

But with your proposed code,  I see in worst case scenario, when all the requests are failed or we see only last request completed successfully,  then both the loop will be executed completely, (i.e. the loop will be executed twice), correct ?